From 715804120fb5fb2565e2b7e687b21fc6467659ff Mon Sep 17 00:00:00 2001
From: Pierre <ppierre@recisio.com>
Date: Sat, 25 Jun 2022 20:33:40 +0200
Subject: [PATCH] Patch for MacOS

---
 src/FFIExtend.php | 26 ++++++++++++++++----------
 1 file changed, 16 insertions(+), 10 deletions(-)

diff --git a/src/FFIExtend.php b/src/FFIExtend.php
index d8a9395..920458d 100644
--- a/src/FFIExtend.php
+++ b/src/FFIExtend.php
@@ -135,16 +135,22 @@ class FFIExtend
 
         $this->replaceMacro('HAVE_LONG_DOUBLE_ZEND_FFI_TYPE_LONGDOUBLE', self::$HAVE_LONG_DOUBLE, 'ZEND_FFI_TYPE_LONGDOUBLE,', $code);
 
-        if (strcasecmp(PHP_OS_FAMILY, 'Windows') === 0) {
-            $this->replaceMacro('ZEND_FASTCALL', 1, '__vectorcall', $code);
-            $phpDll = $this->findPhpDll();
-            self::$ffi = FFI::cdef($code, $phpDll);
-        } else {
-            $this->replaceMacro('ZEND_FASTCALL', 1, '__attribute__((fastcall))', $code);
-            if (PHP_DLL_FILE_PATH) {
-                self::$ffi = FFI::cdef($code, PHP_DLL_FILE_PATH);
-            } else {
-                self::$ffi = FFI::cdef($code);
+        $fastCallMacroPossibleValue = [
+            '__vectorcall',
+            '__attribute__((fastcall))',
+            ''
+        ];
+        $phpDll = strcasecmp(PHP_OS_FAMILY, 'Windows') === 0 ? $this->findPhpDll() : (PHP_DLL_FILE_PATH ?: null);
+        foreach ($fastCallMacroPossibleValue as $possibleValue) {
+            $currentCode = $code;
+            $this->replaceMacro('ZEND_FASTCALL', 1, $possibleValue, $currentCode);
+            try {
+                self::$ffi = FFI::cdef($currentCode, $phpDll);
+                break;
+            } catch (FFI\ParserException $e) {
+                if ($possibleValue === $fastCallMacroPossibleValue[array_key_last($fastCallMacroPossibleValue)]) {
+                    throw $e;
+                }
             }
         }
 
-- 
2.32.1 (Apple Git-133)

